[Analysis] Remove global state from `PluginInline{Advisor,Order}Analysis`.

michele-scandale opened the pull request:
The plugin analysis for `InlineAdvisor` and `InlineOrder` currently relies on shared global state to keep track if the analysis is available.
This causes issues when pipelines using plugins and pipelines not using plugins are run in the same process.
The shared global state can be easily replaced by checking in the given instance of `ModuleAnalysisManager` if the plugin analysis has been registered.

nikic reviewed (approved):
LGTM, being able to directly query whether an analysis was registered sounds reasonable to me. @aeubanks might want to take a look.

aeubanks commented:
I agree that global state breaks a bunch of cases that we care about, but using the pass manager infrastructure to decide whether or not to use the default vs ML inline advisor seems like a hack to avoid having to create the proper APIs. IMO each inliner instance should decide if it's going to use the ML inliner or not, not each pipeline. Thoughts?

michele-scandale commented:
> I agree that global state breaks a bunch of cases that we care about, but using the pass manager infrastructure to decide whether or not to use the default vs ML inline advisor seems like a hack to avoid having to create the proper APIs. 

I believe the proposed change is in line with the original change that introduced the `PluginInlineAdvisorAnalysis` where the plugin behavior takes precedence on the default behavior.
The intention of the proposed change was to tackle the issue of relying on a shared global state.

> IMO each inliner instance should decide if it's going to use the ML inliner or not, not each pipeline. Thoughts?

Today the `InlineAdvisorMode` (specified when instantiating the `ModuleInlinerPass` and `ModuleInlineWrapperPass`) is a global option for the default pipelines (`-enable-ml-inliner`). Whether the plugin is used or not doesn't seem to depend on this option -- i.e. the plugin takes precedence in all the cases -- see https://github.com/llvm/llvm-project/pull/114615/files#diff-ece9bad9424a20a7e1a498a628f00eb3eb3a807fceef8eda927885d0a6591802R207, so I'm not sure how this is connected to the proposed change.

aeubanks commented:
I see some comments at https://github.com/llvm/llvm-project/blob/5005f8d2486d6eec7b2b8ae04f49e8a87ebf4bf6/llvm/include/llvm/Analysis/InlineAdvisor.h#L241 about adding a pass to register an analysis. Is there any reason we can't just register the analysis directly with [`PassBuilder::registerAnalysisRegistrationCallback`](https://github.com/llvm/llvm-project/blob/5005f8d2486d6eec7b2b8ae04f49e8a87ebf4bf6/llvm/include/llvm/Passes/PassBuilder.h#L548) in `llvmGetPassPluginInfo`? that seems like it'd solve your issue IIUC?

michele-scandale commented:
> I see some comments at
> 
> https://github.com/llvm/llvm-project/blob/5005f8d2486d6eec7b2b8ae04f49e8a87ebf4bf6/llvm/include/llvm/Analysis/InlineAdvisor.h#L241
> 
> about adding a pass to register an analysis. Is there any reason we can't just register the analysis directly with [`PassBuilder::registerAnalysisRegistrationCallback`](https://github.com/llvm/llvm-project/blob/5005f8d2486d6eec7b2b8ae04f49e8a87ebf4bf6/llvm/include/llvm/Passes/PassBuilder.h#L548) in `llvmGetPassPluginInfo`? that seems like it'd solve your issue IIUC?

In the context I'm working on there are multiple independent optimization pipelines in the same process for different purposes (each pipeline has its own analysis manager instances): for some of them this exact mechanism is being used (due to a static plugin being registered) to customize the inline advisor, and that end up breaking the other pipelines because of the shared global state.

aeubanks commented:
> This causes issues when pipelines using plugins and pipelines not using plugins are run in the same process.

I think I'm not understanding how

```
llvmGetPassPluginInfo() {
 return {
   [](PassBuilder &PB) {
             PB.registerPipelineStartEPCallback(
                 [](ModulePassManager &MPM, OptimizationLevel Level) {
                   MPM.addPass(DefaultDynamicAdvisor());
                 });
           }};
 }
```

with this patch applied is any different from just

```
llvmGetPassPluginInfo() {
 return {
   [](PassBuilder &PB) {
             PB.registerAnalysisRegistrationCallback(
                 [](ModuleAnalysisManager & MAM) {
                   PluginInlineAdvisorAnalysis PA(defaultAdvisorFactory);
                   MAM.registerPass([&] { return PA; });
                 });
           }};
 }
```

with the `DefaultDynamicAdvisor` pass removed. `registerAnalysisRegistrationCallback` seems like a cleaner way of registering the inline advisor unless I'm not understanding some other detail of `DefaultDynamicAdvisor`

michele-scandale commented:
> > This causes issues when pipelines using plugins and pipelines not using plugins are run in the same process.
> 
> I think I'm not understanding how
> 
> ```
> llvmGetPassPluginInfo() {
>  return {
>    [](PassBuilder &PB) {
>              PB.registerPipelineStartEPCallback(
>                  [](ModulePassManager &MPM, OptimizationLevel Level) {
>                    MPM.addPass(DefaultDynamicAdvisor());
>                  });
>            }};
>  }
> ```
> 
> with this patch applied is any different from just
> 
> ```
> llvmGetPassPluginInfo() {
>  return {
>    [](PassBuilder &PB) {
>              PB.registerAnalysisRegistrationCallback(
>                  [](ModuleAnalysisManager & MAM) {
>                    PluginInlineAdvisorAnalysis PA(defaultAdvisorFactory);
>                    MAM.registerPass([&] { return PA; });
>                  });
>            }};
>  }
> ```
> 
> with the `DefaultDynamicAdvisor` pass removed. `registerAnalysisRegistrationCallback` seems like a cleaner way of registering the inline advisor unless I'm not understanding some other detail of `DefaultDynamicAdvisor`

The problem is on the `InlineAdvisor` side when the `PluginInlineAdvisorAnalysis ` is being queried:
```
bool InlineAdvisorAnalysis::Result::tryCreate(
    InlineParams Params, InliningAdvisorMode Mode,
    const ReplayInlinerSettings &ReplaySettings, InlineContext IC) {
  auto &FAM = MAM.getResult<FunctionAnalysisManagerModuleProxy>(M).getManager();
  if (PluginInlineAdvisorAnalysis::HasBeenRegistered) { // <---- this is a shared global variable!
    auto &DA = MAM.getResult<PluginInlineAdvisorAnalysis>(M);
    Advisor.reset(DA.Factory(M, FAM, Params, IC));
    return !!Advisor;
  }
```

the `HasBeenRegistered` global variable  is being set to `true` in the constructor of `PluginInlineAdvisorAnalysis` -- which currently happen when  the `DefaultDynamicAdvisor::run` is executed, and with your change it will happen when the analysis are registered.
In both cases once someone somewhere does that the `HasBeenRegistered` applies globally, and that will affect whatever other pipelines where plugins are not being loaded.

aeubanks commented:
thanks for pointing out that specific piece of code, I can't come up with a better way of doing that so this PR lgtm.

please update the comments [here](https://github.com/llvm/llvm-project/blob/6f5a145aebee4a925da28d409d69ec7f4ea19f40/llvm/include/llvm/IR/PassManager.h#L460) in this PR, removing the part where it says there's no way to query if an analysis is registered

I'd also request that the comments [here](https://github.com/llvm/llvm-project/blob/5005f8d2486d6eec7b2b8ae04f49e8a87ebf4bf6/llvm/include/llvm/Analysis/InlineAdvisor.h#L241) be updated to use `PassBuilder::registerAnalysisRegistrationCallback` instead of adding a pass that registers the analysis (in a separate PR)

aeubanks commented on llvm/include/llvm/IR/PassManager.h:
```diff
@@ -456,13 +461,6 @@ template <typename IRUnitT, typename... ExtraArgTs> class AnalysisManager {
-  ///
-  /// (Note: Although the return value of this function indicates whether or not
-  /// an analysis was previously registered, there intentionally isn't a way to
-  /// query this directly.  Instead, you should just register all the analyses
```
please keep the bit about just registering analyses without checking if they're already registered

aeubanks replied:
Something like "Although the return value of this function indicates whether or not
  /// an analysis was previously registered, you should just register all the analyses..."

michele-scandale committed:
[Analysis] Remove global state from `PluginInline{Advisor,Order}Analysis`.

The plugin analysis for `InlineAdvisor` and `InlineOrder` currently
relies on shared global state to keep track if the analysis is
available.
This causes issues when pipelines using plugins and pipelines not using
plugins are run in the same process.
The shared global state can be easily replaced by checking in the given
instance of `ModuleAnalysisManager` if the plugin analysis has been
registered.

michele-scandale commented:
@aeubanks Thanks for merging the PR. I would have done that myself today.

> I'd also request that the comments [here](https://github.com/llvm/llvm-project/blob/5005f8d2486d6eec7b2b8ae04f49e8a87ebf4bf6/llvm/include/llvm/Analysis/InlineAdvisor.h#L241) be updated to use PassBuilder::registerAnalysisRegistrationCallback instead of adding a pass that registers the analysis (in a separate PR)

I'll prepare a separate PR for this part.
